Update city_schema.json

hbruch opened the pull request:
This PR fixes #219 by updating city_schema.json.

Note that nevertheless, some current city responses are not valid according to this schema.

Especially Aarhus and Koeln currently report states as "unknown", which IMHO should be reported as "nodata". Further, Koeln reports free=-1 for some parkings. As this seems to represent an unknown value. If true, it should rather be left out, as it is not mandatory.

Note further, that this schema uses some anyOf-type-validations, to allow currently used practice to specify null values for values required by the former schema, but apparently unknown. IMHO null should not be allowed for required properties, but that would result in rework for some data providers.

Finally, this schema s rather strict by not allowing additionalProperties. This kind of blocks experimental features, but uncovers properties not yet specified in the schema. Probably, additionalProperties should better be set to true.

hbruch commented:
Looks like the latest feedparser makes use of python 3.6 string formatting syntax (f'), which fails the build. I guess that either feedparser version should be frozen at an earlier version in requirements.txt or minimum compatibility raised to 3.6.

jklmnn commented:
> Looks like the latest feedparser makes use of python 3.6 string formatting syntax (f'), which fails the build. I guess that either feedparser version should be frozen at an earlier version in requirements.txt or minimum compatibility raised to 3.6.

Thanks for looking at this issue. I think we can safely drop Python 3.5 since even Debian Stable (which we use) has already Python 3.7. (see https://github.com/offenesdresden/ParkAPI/issues/222). Do you have any requirement in this regard? If you need Python 3.6 I would still include it otherwise I'd make Python 3.7 the minimum version.

jklmnn commented on schema/city_schema.json:
```diff
@@ -1,67 +1,128 @@
+              }
+            ]
+          },
+          "state": {
+            "enum": [
+              "open",
+              "nodata",
+              "closed",
+              "many",
+              "few",
+              "full"
+            ]
+          },
```
> Especially Aarhus and Koeln currently report states as "unknown", which IMHO should be reported as "nodata".

`unknown` and `nodata` have two dedicated meanings. `unknown` is the complement to `open` and `closed` which means that even if we have data about the number of free lots we cannot safely tell if the lot is open or closed (some cities do not provide this information). In this case the apps would still show it as open since this is usually the case. `nodata` means we do not have data about the number of free lots.

What are the definitions of `many`, `few` and `full`? Is this only used if there is no more detailed data available? Otherwise I'd say this information is included in the number of free and total lots. Another problem is that these states are not supported by any app yet (although this might be one that can be solved without too much effort).

hbruch replied:
We implemented these states for Herrrenberg, but I'm fine with using state only for open/closed or unkown, if no information is known. Isn't `nodata` redundant for having no count information and additionally hide open/close then?

jklmnn replied:
> Isn't `nodata` redundant for having no count information and additionally hide open/close then?

I'm not sure if I understand the question correctly. `nodata` implies that there is no live data or not enough to be useful. It implies `unknown`. The desired handling for applications is that if `nodata` is set the application simply shows the parking lot but only with static data (location, type, total count, etc.). The case that we know if a lot is open or closed but we don't have data on how full it is cannot be modeled currently.

hbruch replied:
Is it really necessary to have a state `nodata`? Couldn't this implicitly deduced when no `free` value is given? Currently, there are instances, which return free=-1 if no data is available. Separating opening state and realtime availability in two different property allows to communicate open/closed state without knowing current occupancy.

My proposition is to keep only open/closed/unknown and remove nodata as well as our introduced few/many/full values.

jklmnn replied:
We have to distinguish between what is logically useful (where you are completely correct) and what we need for compatibility. My current motivation is to keep the latter (where we also need `nodata`). We are aware that the current API has issues but we don't want to "fix" it in a way that would break compatibility since it has some drawbacks that cannot be fixed at all with this architecture. We have outlined the main problems with the current API in https://github.com/offenesdresden/ParkAPI/issues/130.
Further I know that removing `nodata` will break some of our applications. Also I have noticed that there are a few third party implementations that use our API where I don't know how exactly they are implemented and I don't want to break them either with such a small change.

hbruch replied:
I'm fine with leaving `nodata` in the schema for know, though I think that clients should be capable to handle a missing free value as well. I'll adapt this PR (removing the frozen feed parser version as well) after #222 has been fixed.

hbruch replied:
Now fixed

hbruch commented:
No objections to moving to python 3.7 as minimum requirement. I'll open an issue to gather what needs to be done.

hbruch commented:
I resolve the state related change requests, switched additonalProperties to true, to allow custom extensions and updated the README.md example to reflect the schema.

jklmnn commented:
Fixed by https://github.com/offenesdresden/ParkAPI/commit/3f2eb3729965812a096c2ca93bee376689204b93.
